lib/kubernetes-deploy/renderer.rb

```diff
@@ -0,0 +1,99 @@
+# frozen_string_literal: true
+
+require 'erb'
+require 'securerandom'
+require 'yaml'
+require 'json'
+
+module KubernetesDeploy
+  class Renderer
+    def initialize(current_sha:, template_dir:, logger:, bindings: {})
+      @current_sha = current_sha
+      @template_dir = template_dir
+      @partials_dirs =
+        %w(partials ../partials).map { |d| File.expand_path(File.join(@template_dir, d)) }
+      @logger = logger
+      @bindings = bindings
+      # Max length of podname is only 63chars so try to save some room by truncating sha to 8 chars
+      @id = current_sha[0...8] + "-#{SecureRandom.hex(4)}" if current_sha
+    end
+
+    def render_template(filename, raw_template)
+      return raw_template unless File.extname(filename) == ".erb"
+
+      erb_binding = TemplateContext.new(self).template_binding
+      bind_template_variables(erb_binding, template_variables)
+
+      ERB.new(raw_template, nil, '-').result(erb_binding)
+    rescue StandardError => err
```

Comment:
Gave the feature another spin, including with some invalid templates, and noticed that rescuing `StandardError` both here and in `render_partial` is leading to some confusing error output when a partial is invalid (first screenshot) or includes a missing partial (second screenshot):

![image](https://user-images.githubusercontent.com/4789493/36461971-974d63e2-168e-11e8-9584-350d5874c973.png)

![image](https://user-images.githubusercontent.com/4789493/36463779-02cc0cee-169a-11e8-92ba-889027a39f13.png)


Let's not only fix that, but also provide the render sequence to help people debug. Here's one way to do that: https://github.com/Shopify/kubernetes-deploy/compare/9cf798f...partial_review3

Since logging is our main UI in this project, I'd also like to see some integration tests with thorough assertions for those error cases. That commit also shows what I have in mind there. For reference, here's the "after" output:

![image](https://user-images.githubusercontent.com/4789493/36463754-cf346a34-1699-11e8-923b-8f1b49a09762.png)

![image](https://user-images.githubusercontent.com/4789493/36463761-e2a49eea-1699-11e8-9b95-1be307f5fa45.png)